source/extensions/filters/http/proto_message_scrubbing/filter.cc

```diff
@@ -0,0 +1,434 @@
+        *encoder_callbacks_);
+    return Envoy::Http::FilterHeadersStatus::Continue;
+  }
+
+  if (!extractor_) {
+    return Envoy::Http::FilterHeadersStatus::Continue;
+  }
+
+  // Create response_msg_converter to convert response body.
+  auto cord_message_data_factory = std::make_unique<CreateMessageDataFunc>(
+      []() { return std::make_unique<Protobuf::field_extraction::CordMessageData>(); });
+
+  response_msg_converter_ = std::make_unique<MessageConverter>(
+      std::move(cord_message_data_factory), encoder_callbacks_->encoderBufferLimit());
+
+  return Http::FilterHeadersStatus::StopIteration;
+}
+
+Envoy::Http::FilterDataStatus Filter::encodeData(Envoy::Buffer::Instance& data, bool end_stream) {
+  ENVOY_STREAM_LOG(debug, "encodeData: data size={} end_stream={}", *decoder_callbacks_,
+                   data.length(), end_stream);
+
+  if (!response_msg_converter_ || !extractor_) {
+    return Envoy::Http::FilterDataStatus::Continue;
+  }
+
+  if (auto status = handleEncodeData(data, end_stream); !status.got_messages) {
+    return status.filter_status;
+  }
+
+  return Envoy::Http::FilterDataStatus::Continue;
+}
+
+Filter::HandleDataStatus Filter::handleEncodeData(Envoy::Buffer::Instance& data, bool end_stream) {
+  RELEASE_ASSERT(extractor_ && response_msg_converter_,
+                 "`extractor_` and response_msg_converter_ both should be "
+                 "initiated when scrubbing proto messages");
+
+  auto buffering = response_msg_converter_->accumulateMessages(data, end_stream);
+
+  if (!buffering.ok()) {
+    const absl::Status& status = buffering.status();
+    rejectResponse(status.raw_code(), status.message(),
+                   generateRcDetails(kRcDetailFilterProtoMessageScrubbing,
+                                     absl::StatusCodeToString(status.code()),
+                                     kRcDetailErrorResponseBufferConversion));
+    return HandleDataStatus(Envoy::Http::FilterDataStatus::StopIterationNoBuffer);
+  }
+
+  if (buffering->empty()) {
+    ENVOY_STREAM_LOG(debug, "not a complete msg", *encoder_callbacks_);
+    // Not a complete message.
+    return HandleDataStatus(Envoy::Http::FilterDataStatus::StopIterationNoBuffer);
+  }
+
+  // Buffering returns a list of messages.
+  for (size_t msg_idx = 0; msg_idx < buffering->size(); ++msg_idx) {
+    std::unique_ptr<StreamMessage> stream_message = std::move(buffering->at(msg_idx));
+
+    // The converter returns an empty stream_message for the last empty
+    // buffer.
+    if (stream_message->message() == nullptr) {
+      DCHECK(end_stream);
+      DCHECK(stream_message->isFinalMessage());
+      // This is the last one in the vector.
+      DCHECK(msg_idx == buffering->size() - 1);
+      continue;
+    }
+
+    // Set the response_scrubbing_done_ to true since we have received a message for scrubbing.
+    response_scrubbing_done_ = true;
+
+    extractor_->processResponse(*stream_message->message());
+
+    if (end_stream) {
```

Comment:
What if there's trailers?

Reply:
Good point. We do not need to check `end_stream` here. Message Converter handles that logic. Removed the if condition from request/response path.

Also, added a test case with `end_stream=false` to indicate trailers' presence.

Thanks for noticing!